### ISSUE TITLE
Concourse & Vault Integration Error: resource config creds evaluation: undefined vars:

### ISSUE THREAD

**Issue by ChrisJBurns** [concourse/concourse#8175]
## Summary
I'm using Vault and Concourse Helm charts. I have managed to get both working successfully, however, there are times specifically when the `concourse-web` pod is restarted for whatever reason, that when it restarts, I get the `resource config creds evaluation: undefined vars:` error when trying to retrieve secrets from Vault. It will be working completely fine up until the restart of the concourse-web pod.

Does anyone know why this keeps happening? I am using the `approle` auth backend, and adding it to the correct secret values - as I've managed to get it working fine. But it's getting to a point where countless hours have been lost having to delete the approle in vault, recreate it, get a new credentials to add to the concourse secret. Even this is temperamental as sometimes it resolves and sometimes it doesn't. Is there any official way of doing this? Or is the case valid where every time the concourse-web pod restarts that it completely severs ties with all working approle's?

My very basic assumption is that the logging in with the `approle` details aren't happening? I can use the Vault app role credentials fine via the API, and I see the entity appear in Vault. But for some reason, with Concourse, it sometimes doesn't happen

## Steps to reproduce
- Install Vault and Concourse via Helm (include the preconfigured `roleId` and `secretId` in the Concourse config via the Helm chart)
- Once both are installed, init and unseal Vault and create the `approle` for Concourse using the `roleId` and `secretId` mentioned above.
- Wait for concourse to use it, and it doesn't. In the pipelines, when I try to access the secrets I get the error described in title. For some reason, its not logging in with the configured credentials.

**Comment by ChrisJBurns** [comment]
Hi,

Not really, I've been trying to solve this problem for a couple of months now and nothing seems to really work.

Right now for example, I have Concourse configured with an `useAuthParam` of both a `roleId` and `secretId`. 
When I create the `approle` in Vault, I use terraform that does the following

```yaml
resource "vault_approle_auth_backend_role" "concourse_app_role" {
  backend        = vault_auth_backend.approle.path
  role_name      = "concourse"
  token_policies = [ "concourse" ]
  token_no_default_policy = true
  token_period = 3600
  role_id = yamldecode(file("config.yaml"))["vault"]["concourse-approle"]["role-id"]
}

resource "vault_approle_auth_backend_role_secret_id" "concourse_app_role_secret_id" {
  backend   = vault_auth_backend.approle.path
  role_name = vault_approle_auth_backend_role.concourse_app_role.role_name
  secret_id = yamldecode(file("config.yaml"))["vault"]["concourse-approle"]["secret-id"]
}
```

When this runs successfully, I still cannot see Concourse actually work. I also check the entities page in Vault and I can see that Concourse isn't even logging in. So I wonder if there is a config option somewhere where I can force Concourse to retry and authenticate. Currently I have the following Vault config in the concourse Helm chart, but it doesn't really seem to do anything.

```yaml
       vault:
          enabled: true
          url: http://vault.vault:8200
          authBackend: "approle"
          pathPrefix: /secrets
          lookupTemplates: [ "/{{.Team}}/{{.Secret}}" ]
          useAuthParam: true
          loginTimeout: 30m
          queryTimeout: 30m
````

**Comment by ChrisJBurns** [comment]
In case it's worth adding, Vault is installed first, then Concourse. However Concourse is installed with the `roleId` and `secretId`. It is only after both Concourse and Vault are both installed, do I go in and create the AppRole within Vault with the same `roleId` and `secretId` values as was specified in the Concourse config.

So maybe Concourse is trying to log in but finds that Vault isn't really functional yet? But I thought the reauth logic would have retried until it worked.

**Comment by xtremerui** [comment]
If you have access to concourse web node log, you should see some logs related to vault creds manager, especially when web node restarts.

**Comment by ChrisJBurns** [comment]
Can confirm, I just restarted the `concourse-web` deployment, and I see no mention of the word `vault` in the logs. The error in issue description has come up though.

```
{"timestamp":"2022-03-17T19:22:39.141931956Z","level":"info","source":"atc","message":"atc.tracker-imb.run.finish.errored","data":{"build":"check","error":"resource config creds evaluation: undefined vars: shared.concourse_private_key","pipeline":"bootstrap","pre_build_id":20,"resource":"test-project","session":"23.20.4","team":"dev-team"}}
````

**Comment by ChrisJBurns** [comment]
Hmmm, I wonder if `{"timestamp":"2022-03-17T19:26:25.866889948Z","level":"info","source":"atc","message":"atc.credential-manager.configured credentials manager","data":{"name":"kubernetes","session":"7"}} ` has anything to do with it?

Doing a bit of Googling and come across this: https://github.com/concourse/concourse/issues/2981.

Funny enough, I've just set that `CONCOURSE_KUBERNETES_IN_CLUSTER` value to `false` in my `concourse-web` deployment and I can confirm that the `concourse-web` node instead logged out `{"timestamp":"2022-03-17T19:37:12.277735893Z","level":"info","source":"atc","message":"atc.credential-manager.configured credentials manager","data":{"name":"vault","session":"7"}}`.

In this case, I think that may have been the issue. Let me just double check by destroying all of the pipelines and Vault configurations and redeploy and see if that forces Concourse to use Vault.

**Comment by ChrisJBurns** [comment]
Yes, that seemed to have fixed it. So it seemed that without the `CONCOURSE_KUBERNETES_IN_CLUSTER` flag set to false, for some reason Concourse defaults to the Kubernetes secrets manager instead of Vault (which is what I have enabled). So on top of the Vault enablement in the Concourse Helm Chart config, I have to add that flag as well.

**Comment by ChrisJBurns** [comment]
Can confirm that setting the following worked.

```yaml
    concourse:
      web:
        kubernetes:
          enabled: false
```

So, if people are using Vault as a credential manager then, they may have to also set the `kubernetes.enabled` flag to false as Concourse will try and use it as well as Vault - therefore causing problems when retrieving secrets.

So, perhaps a bit of code is added to Concourse to error when multiple secrets backends are configured? Or even make it known on the [documentation for Vault integration](https://concourse-ci.org/vault-credential-manager.html) (and possibly other backends if it also happens there) that the extra config will need to be added to ensure there's no conflict.

**Comment by factorypreset** [comment]
We've been seeing this same issue recently with AWS ssm as the secrets backend, and I can confirm that this same configuration also fixed our issue. So thanks for figuring this out!

**Comment by ChrisJBurns** [comment]
Glad my spamming of comments helped :) 

I've also realised that, it is actually quite temperamental. I've found that _sometimes_ the `concourse-web` pod will actually start with Vault as it's secrets backend, and then sometimes it will do the Kubernetes one. Although I can definitely say that it starts up with a Kubernetes secrets backend more times than it does with Vault. Out of 10 restarts, I'm probably going to go with 8/9 of them being Kubernetes and 1/2 of them being Vault.

The only thing I know of at the moment, that gets Vault to be the secrets backend consistently is by applying the above fix.
